common/src/main/java/io/pravega/common/util/OrderedItemProcessor.java

```diff
@@ -0,0 +1,253 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import io.pravega.common.ExceptionHelpers;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.FutureHelpers;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.val;
+
+/**
+ * Processes items in order, subject to capacity constraints.
+ */
+@ThreadSafe
+public class OrderedItemProcessor<ItemType, ResultType> implements AutoCloseable {
+    //region members
+
+    private static final int CLOSE_TIMEOUT_MILLIS = 60 * 1000;
+    private final int capacity;
+    @GuardedBy("processingLock")
+    private final Function<ItemType, CompletableFuture<ResultType>> processor;
+    @GuardedBy("stateLock")
+    private final Deque<QueueItem> pendingItems;
+    private final Executor executor;
+
+    /**
+     * Guards access to the OrderedItemProcessor's internal state (counts and queues).
+     */
+    private final Object stateLock = new Object();
+
+    /**
+     * Guards access to the individual item's processors. No two invocations of the processor may run at the same time.
+     */
+    private final Object processingLock = new Object();
+    @GuardedBy("stateLock")
+    private int activeCount;
+    @GuardedBy("stateLock")
+    private boolean closed;
+    @GuardedBy("stateLock")
+    private ReusableLatch emptyNotifier;
```

Comment:
I wanted something simpler and lower-level than a CompletableFuture. I looked at CF's implementation and it looked too heavyweight for what I needed. All I needed here was some sort of basic signaling.